### PR TITLE
[exporter/coralogix] Initial implementation

### PR DESCRIPTION
**Pull request by ofirshmuel**
**Description:** <Describe what has changed.>

Second PR of Coralogix exporter , the client connect to Coralogix and sending traces

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** <Describe what testing was performed and which tests were added.>

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

**Review by oded-dd — APPROVED** [review]

File: exporter/coralogixexporter/client.go
Comment (jpkrohling) [inline exporter/coralogixexporter/client.go:36]: Should this be exported?

File: exporter/coralogixexporter/client.go
Comment (jpkrohling) [inline exporter/coralogixexporter/client.go:53]: I would really avoid dealing with env vars here. Users can use env vars in the config, and this PR causes this component to deviate from other components.

File: exporter/coralogixexporter/client.go
Comment (jpkrohling) [inline exporter/coralogixexporter/client.go:69]: The typical pattern is to add `configgrpc.GRPCClientSettings` directly to the config object so that users can control most of the aspects of this here, including the TLS parts. Like this:

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/eacfde3fcbd46ba60a6db0e9a41977390c4883bd/exporter/skywalkingexporter/config.go#L28

File: exporter/coralogixexporter/client.go
Comment (jpkrohling) [inline exporter/coralogixexporter/client.go:99]: Please, remove this print

File: exporter/coralogixexporter/client.go
Comment (jpkrohling) [inline exporter/coralogixexporter/client.go:103]: This doesn't belong here either :-)

File: exporter/coralogixexporter/client.go
Comment (jpkrohling) [inline exporter/coralogixexporter/client.go:104]: We just got a PR merged in the core that should take care of this logging: open-telemetry/opentelemetry-collector#4501

File: exporter/coralogixexporter/client.go
Comment (jpkrohling) [inline exporter/coralogixexporter/client.go:111]: All prints should be removed.

File: exporter/coralogixexporter/client.go
Comment (jpkrohling) [inline exporter/coralogixexporter/client.go:107]: This should accept a `context.Context`, and pass it as part of the outgoing context.

File: exporter/coralogixexporter/client.go
Comment (jpkrohling) [inline exporter/coralogixexporter/client.go:114]: Instead of the todo, use the context I mentioned earlier.

File: exporter/coralogixexporter/config_test.go
Comment (jpkrohling) [inline exporter/coralogixexporter/config_test.go:56]: It's probably better to remove the logging from the test at this point.

File: exporter/coralogixexporter/exporter.go
Comment (jpkrohling) [inline exporter/coralogixexporter/exporter.go:42]: Pass the context you received.

File: exporter/coralogixexporter/exporter.go
Comment (jpkrohling) [inline exporter/coralogixexporter/exporter.go:26]: Should this be exported?

File: exporter/coralogixexporter/exporter.go
Comment (jpkrohling) [inline exporter/coralogixexporter/exporter.go:28]: You might want to store a telemetrysettings here instead, like this:

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/eacfde3fcbd46ba60a6db0e9a41977390c4883bd/exporter/skywalkingexporter/skywalking.go#L46

It contains a logger, a tracer, and a meter.

File: exporter/coralogixexporter/exporter.go
Comment (jpkrohling) [inline exporter/coralogixexporter/exporter.go:36]: If you do decide to use the telemetry settings, it's available as `params.TelemetrySettings`

File: exporter/coralogixexporter/client.go
Comment (ofirshmuel) [reply exporter/coralogixexporter/client.go:36]: No, I will change this. the access to coralogixClient is internally, only from coralogixexporter package

File: exporter/coralogixexporter/exporter.go
Comment (ofirshmuel) [reply exporter/coralogixexporter/exporter.go:26]: No, I will change this. the access to coralogixExporter is internally, only from coralogixexporter package

File: exporter/coralogixexporter/exporter.go
Comment (ofirshmuel) [reply exporter/coralogixexporter/exporter.go:28]: Thanks, but in this time we not use component.TelemetrySettings

File: exporter/coralogixexporter/client.go
Comment (ofirshmuel) [reply exporter/coralogixexporter/client.go:69]: I will add this config in the next commit

File: exporter/coralogixexporter/client.go
Comment (ofirshmuel) [reply exporter/coralogixexporter/client.go:36]: but after I changed this to "coralogixClient", the linter throw exception on function NewCoralogixClient declaration in client.go.
`unexported-return: exported func NewCoralogixClient returns unexported type *coralogixexporter.coralogixClient, which can be annoying to use (revive)`.
I can't declare this function with return value "*coralogixClient"

Comment (jpkrohling) [reply exporter/coralogixexporter/client.go:32]: Because `NewCoralogixClient` also likely needs to kept private.

File: exporter/coralogixexporter/config.go
Comment (jpkrohling) [inline exporter/coralogixexporter/config.go:77]: What happens if users have specified headers themselves? You'd be overriding them here.

File: exporter/coralogixexporter/config.go
Comment (jpkrohling) [inline exporter/coralogixexporter/config.go:38]: The gRPC client settings object already has an Endpoint, use that.

File: exporter/coralogixexporter/factory.go
Comment (jpkrohling) [inline exporter/coralogixexporter/factory.go:49]: Validate is called automatically by the collector itself before your exporter is created.

File: exporter/coralogixexporter/config.go
Comment (jpkrohling) [inline exporter/coralogixexporter/config.go:69]: You might want to take the opportunity and fix the error message as well.

File: exporter/coralogixexporter/client.go
Comment (jpkrohling) [inline exporter/coralogixexporter/client.go:74]: If you just return the error and bubble it up, your users can take advantage of the retry logic that can be configured as part of their pipeline. If you don't return the error, the collector will assume this worked and won't retry.

File: exporter/coralogixexporter/client.go
Comment (ofirshmuel) [reply exporter/coralogixexporter/client.go:32]: Thanks, I will change this

File: exporter/coralogixexporter/config.go
Comment (ofirshmuel) [reply exporter/coralogixexporter/config.go:77]: Yes, I want to control about request header, I want to override the users.

File: exporter/coralogixexporter/factory.go
Comment (ofirshmuel) [reply exporter/coralogixexporter/factory.go:49]: Thanks, I will remove this code line

File: exporter/coralogixexporter/config.go
Comment (ofirshmuel) [reply exporter/coralogixexporter/config.go:69]: Thanks, I will fix the error message

File: exporter/coralogixexporter/config.go
Comment (ofirshmuel) [reply exporter/coralogixexporter/config.go:38]: I want to stay with Endpoint configuration name to make the config file easier to use.
After the users put the endpoint we modify the GRPCClientSettings.
Can you explain why to remove "endpoint" ?

File: exporter/coralogixexporter/config.go
Comment (jpkrohling) [reply exporter/coralogixexporter/config.go:90]: If your users are specifying any other HTTP headers, perhaps because a hop between them and you requires it, you'd be dropping their values without warning. Be nice to your users, if you really need to drop the headers they explicitly added, at least tell them you are removing it.

File: exporter/coralogixexporter/config.go
Comment (jpkrohling) [reply exporter/coralogixexporter/config.go:38]: There's no difference for your users, as you are squashing the configuration.

**Comment by ofirshmuel** [comment]
@jpkrohling any update on this PR?

File: exporter/coralogixexporter/config.go
Comment (ofirshmuel) [reply exporter/coralogixexporter/config.go:90]: Thanks, I will not override any HTTP header, only the specific header keys

File: exporter/coralogixexporter/config.go
Comment (ofirshmuel) [reply exporter/coralogixexporter/config.go:38]: Thanks, I will use it

**Comment by jpkrohling** [comment]
While you are working on the PR comments, could you also please go over #6926 and see if there's anything missing?

**Comment by ofirshmuel** [comment]
@jpkrohling  I run over your comments, it's seems to include everything require in the first PR, let me know if anything else is missing.

File: exporter/coralogixexporter/client.go
Comment (jpkrohling) [inline exporter/coralogixexporter/client.go:77]: This shouldn't be here. Also, please address the comment a couple of lines above this one, about returning errors that could be retried later.

File: exporter/coralogixexporter/client.go
Comment (jpkrohling) [inline exporter/coralogixexporter/client.go:58]: See other exporters how they build the gRPC connection. Basicall, you'd call the GRPCClientSettings.ToDialOptions().

File: exporter/coralogixexporter/client.go
Comment (ofirshmuel) [reply exporter/coralogixexporter/client.go]: Thanks, I will fix it

File: exporter/coralogixexporter/client.go
Comment (ofirshmuel) [reply exporter/coralogixexporter/client.go]: Thanks, I will change this method in the next commit.

**Comment by ofirshmuel** [comment]
@jpkrohling any updates?

**Comment by jpkrohling** [comment]
I'll review this again shortly.

File: exporter/coralogixexporter/client.go
Comment (jpkrohling) [reply exporter/coralogixexporter/client.go:77]: Are you going to address this? It's your component, you can decide to not do it, but IMO, it gives your users a poor and inconsistent experience.

File: exporter/coralogixexporter/config.go
Comment (jpkrohling) [inline exporter/coralogixexporter/config.go:55]: I believe I left a comment earlier about the usage of env vars here. Note that it's possible to specify the same component multiple times, so, you wouldn't know to which component this env var relates to. For instance:

```
exporters:
  coralogix/department1:
    application_name: "APP_NAME"
    subsystem_name: "SUBSYSTEM_NAME"
  coralogix/department2:
    application_name: "APP_NAME"
    subsystem_name: "SUBSYSTEM_NAME"
```

When CORALOGIX_ENDPOINT or CORALOGIX_PRIVATE_KEY is specified, should it really be applied to all coralogix exporters in this collector?

File: exporter/coralogixexporter/client.go
Comment (ofirshmuel) [reply exporter/coralogixexporter/client.go:78]: Thanks, I will fix it, the error will bubble up to the users.

File: exporter/coralogixexporter/config.go
Comment (ofirshmuel) [reply exporter/coralogixexporter/config.go:55]: Thanks for a good idea but in Coralogix we support only one application and subsystem per user.
Multiple is not available for us.

File: exporter/coralogixexporter/client.go
Comment (jpkrohling) [inline exporter/coralogixexporter/client.go:70]: When returning the error, don't log it here, as it will be logged elsewhere up the call chain already.

Comment (ofirshmuel) [reply exporter/coralogixexporter/client.go:70]: @jpkrohling I want the user will get an indicative message, in addition to the default error.

File: exporter/coralogixexporter/config.go
Comment (jpkrohling) [reply exporter/coralogixexporter/config.go:55]: @open-telemetry/collector-contrib-approvers I guess I need a second opinion from another approver.

File: exporter/coralogixexporter/client.go
Comment (jpkrohling) [reply exporter/coralogixexporter/client.go:70]: I need a second opinion from another approver here as well (@open-telemetry/collector-contrib-approvers).

**Comment by jpkrohling** [comment]
I'm not very happy with the deviations from the common behavior of the collector, but given that I might be unreasonable in blocking this because of that, I'm asking for another approver to take a look at the contention points.

File: exporter/coralogixexporter/config.go
Comment (Aneurysm9) [reply exporter/coralogixexporter/config.go:55]: The direct use of environment variables is not appropriate.  This should be configured in the same manner as every other configuration item for the component: through the configuration system.  While your application may not support multiple applications or subsystems per user, the collector can support an operator who wishes to configure multiple users to achieve a goal of having multiple applications and/or subsystems.

File: exporter/coralogixexporter/client.go
Comment (Aneurysm9) [reply exporter/coralogixexporter/client.go:70]: If the error is going to be logged further up the call stack but you want to ensure that the logged message contains context regarding where specifically the error was thrown and in what circumstances, that is a situation where the error should be wrapped before it is returned.

Instead of this:

```go
return err
```

do this:

```go
return fmt.Errorf("can't tpanslate to jaeger proto: %w", err)
```

Comment (ofirshmuel) [reply exporter/coralogixexporter/client.go:70]: Thanks, @jpkrohling and @Aneurysm9 I will fix it.

**Comment by bogdandrutu** [comment]
@jpkrohling assigned this to you, since @owais is not an approver anymore.

**Comment by ofirshmuel** [comment]
@owais any updates?

File: exporter/coralogixexporter/config.go
Comment (ofirshmuel) [reply exporter/coralogixexporter/config.go]: @Aneurysm9 @jpkrohling Thanks, I will remove the environments variables support and I will update the readme accordingly.

**Comment by ofirshmuel** [comment]
@jpkrohling please see @bogdandrutu comment, how can we proceed?

**Comment by jpkrohling** [comment]
I'll re-review this soon.

**Comment by ofirshmuel** [comment]
@jpkrohling any updates?

File: exporter/coralogixexporter/README.md
Comment (jpkrohling) [inline exporter/coralogixexporter/README.md:23]: Is this duplicated?

File: exporter/coralogixexporter/config.go
Comment (jpkrohling) [inline exporter/coralogixexporter/config.go:66]: Why do you have the same things twice?

File: exporter/coralogixexporter/example/config-example-1.yaml
Comment (jpkrohling) [inline exporter/coralogixexporter/example/config-example-1.yaml:2]: Why opencensus?

**Comment by jpkrohling** [comment]
Please, include also a change to the changelog.

File: exporter/coralogixexporter/example/config-example-1.yaml
Comment (ofirshmuel) [reply exporter/coralogixexporter/example/config-example-1.yaml:2]: This is an example of the configuration file for metrics.
Opencensus is just an example of a metrics reciver.

File: exporter/coralogixexporter/README.md
Comment (ofirshmuel) [reply exporter/coralogixexporter/README.md:23]: Thanks, I will fix it

File: exporter/coralogixexporter/config.go
Comment (ofirshmuel) [reply exporter/coralogixexporter/config.go:66]: Thanks, I will fix it

File: exporter/coralogixexporter/example/config-example-1.yaml
Comment (jpkrohling) [reply exporter/coralogixexporter/example/config-example-1.yaml:2]: Can't you use OTLP here? OpenCensus will eventually be deprecated.

File: exporter/coralogixexporter/factory.go
Comment (jpkrohling) [inline exporter/coralogixexporter/factory.go:57]: Not for this PR, but a better pattern is to have your exporter have a `start` function, which in turn calls `client.startConnection`. This way, your factory doesn't have to know about the implementation details of the exporter.

**Review by jpkrohling — APPROVED** [review]

**Comment by jpkrohling** [comment]
@Aneurysm9, would you like to give this one another review?

**Comment by ofirshmuel** [comment]
@jpkrohling this is not the last pr, according to the requirements, for new exporter needs 3 PRs.
This is the second, in the next PR I will add the exporter to components.go file.
Any update for this PR?

**Review by Aneurysm9 — APPROVED** [review]
If I understand correctly, this is the Jaeger exporter using a custom proto that includes some extra metadata.  Unfortunate that it can't be sent through headers or the like, as appears to have been done with the `prometheusremotewrite` example, but if this is how it needs to be the implementation looks sane.

File: exporter/coralogixexporter/example/config-example-1.yaml
Comment (ofirshmuel) [reply exporter/coralogixexporter/example/config-example-1.yaml:2]: Ok, I will change it to "OTLP" receiver

File: exporter/coralogixexporter/example/config-example-1.yaml
Comment (ofirshmuel) [reply exporter/coralogixexporter/example/config-example-1.yaml]: In the last commit, I removed the unneeded line and the receiver changed to "OTLP".

**Comment by ofirshmuel** [comment]
@Aneurysm9 In addition to metadata proto, The headers send with GRPCClientSettings. The Coralogix endpoints use these headers.

**Comment by ofirshmuel** [comment]
Hi @Aneurysm9 , I see that the PR is failing on a test, looking at it, it does not seem to be related to the E2E integrations, and not directly to our PR, can you please run it again and merge.
Thanks :)

**Comment by Aneurysm9** [comment]
> Hi @Aneurysm9 , I see that the PR is failing on a test, looking at it, it does not seem to be related to the E2E integrations, and not directly to our PR, can you please run it again and merge. Thanks :)

That workflow fails occasionally, particularly around releases since it makes some assumptions about versions.  I don't think it is an issue.  

I do not have the ability to merge this PR.  @jpkrohling or @codeboten perhaps can help.

**Comment by ofirshmuel** [comment]
@jpkrohling I see that you approved the PR code. The tests failed but it does not seem to be related to the E2E integrations, and not directly to our PR, can you please run it again and merge.

**Comment by jpkrohling** [comment]
Merged, thanks for pinging me!

**Pull request merged**
